tensorrt_llm/_torch/auto_deploy/custom_ops/torch_attention.py

```diff
@@ -163,3 +165,293 @@ def fused_mha_fake(
 ) -> torch.Tensor:
     """Fake Fused MHA+Rope that takes raw input from q, k, v GEMMs."""
     return torch.empty_like(q)
+
+
+def update_kv_cache(
+    key_states: torch.Tensor,
+    value_states: torch.Tensor,
+    k_cache: torch.Tensor,
+    v_cache: torch.Tensor,
+    seq_len: torch.Tensor,  # metadata
+    input_pos: torch.Tensor,  # metadata
+    cache_loc: torch.Tensor,
+    seq_start: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Reference implementation for update kv cache function. Assumes KV cache layout to be [B,S,N,D].
+    This function can be used to build reference attention implementations that use KV cache.
+    """
+
+    for idx in range(seq_len.shape[0]):
+        k_cache[cache_loc[idx], input_pos[idx] : input_pos[idx] + seq_len[idx], :, :] = key_states[
+            seq_start[idx] : seq_start[idx] + seq_len[idx], ...
+        ]
+        v_cache[cache_loc[idx], input_pos[idx] : input_pos[idx] + seq_len[idx], :, :] = (
+            value_states[seq_start[idx] : seq_start[idx] + seq_len[idx], ...]
+        )
+
+
+# Copied from transformers.models.llama.modeling_llama.rotate_half
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+# Copied from transformers.models.llama.modeling_llama.apply_rotary_pos_emb
+@torch.inference_mode()
+def apply_rotary_pos_emb_ds(q, k, cos, sin, position_ids, unsqueeze_dim=1, flattened=True):
```

Comment:
```suggestion
def apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
```

It's not deepseek-specific anymore and the `flattened` argument is not used anymore

Reply:
I see the issue. Let's leave it as is for now and consider a better structure around rope in the future as just discussed in the group meeting.